src/probability/extrealScript.sml

```diff
@@ -9463,6 +9463,268 @@ Proof
+Proof
+    ‘∀f g l. (∀e. MEM e l ⇒ f e ≤ g e) ⇒
+      (FOLDR (λe acc. f e + acc) 0x l ≤ FOLDR (λe acc. g e + acc) 0x l)’
+        suffices_by rw[EXTREAL_SUM_IMAGE_ALT_FOLDR] >>
+    Induct_on ‘l’ >> rw[listTheory.FOLDR] >> irule le_add2 >> simp[]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_ZERO:
+    ∀f. ∑ f (count 0) = 0:extreal
+Proof
+    simp[COUNT_ZERO,EXTREAL_SUM_IMAGE_EMPTY]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_ONE:
+    ∀f. ∑ f (count 1) = f 0:extreal
+Proof
+    simp[COUNT_ONE,EXTREAL_SUM_IMAGE_SING]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_SUC:
+    ∀f n. (∀m. m ≤ n ⇒ f m ≠ −∞) ∨ (∀m. m ≤ n ⇒ f m ≠ +∞) ⇒
+        ∑ f (count (SUC n)) = (∑ f (count n)) + f n:extreal
+Proof
+    rw[] >> ‘count (SUC n) = (count n) ∪ {n}’ by fs[count_def,EXTENSION] >>
+    rw[] >> pop_assum kall_tac >>
+    ‘∑ f (count n ∪ {n}) = ∑ f (count n) + ∑ f {n}’ suffices_by fs[EXTREAL_SUM_IMAGE_SING] >>
+    irule EXTREAL_SUM_IMAGE_DISJOINT_UNION >> simp[]
+QED
+
+(*** EXTREAL_PROD_IMAGE Theorems ***)
+
+Theorem EXTREAL_PROD_IMAGE_NOT_INFTY:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ f x ≠ −∞ ∧ f x ≠ +∞) ⇒ ∏ f s ≠ −∞ ∧ ∏ f s ≠ +∞
+Proof
+    strip_tac >> simp[Once $ GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >> CONJ_TAC
+    >- simp[EXTREAL_PROD_IMAGE_EMPTY,SYM normal_1] >>
+    NTAC 5 strip_tac >> fs[EXTREAL_PROD_IMAGE_PROPERTY,DELETE_NON_ELEMENT_RWT] >>
+    Cases_on ‘f e’ >> Cases_on ‘∏ f s’ >> rfs[extreal_mul_def]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_NORMAL:
+    ∀f s. FINITE s ⇒ ∏ (λx. Normal (f x)) s = Normal (∏ f s)
+Proof
+    strip_tac >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,REAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT,extreal_mul_def,normal_1]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_0:
+    ∀f s. FINITE s ∧ (∃x. x ∈ s ∧ f x = 0) ⇒ ∏ f s = 0
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >- fs[] >>
+    DISJ2_TAC >> first_x_assum irule >> qexists_tac ‘x’ >> simp[]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_1:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ f x = 1) ⇒ ∏ f s = 1
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_ONE:
+    ∀s. FINITE s ⇒ ∏ (λx. 1) s = 1x
+Proof
+    Induct_on ‘s’ >> simp[EXTREAL_PROD_IMAGE_EMPTY,EXTREAL_PROD_IMAGE_PROPERTY,DELETE_NON_ELEMENT_RWT]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_POS:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ 0 ≤ f x) ⇒ 0 ≤ ∏ f s
+Proof
+    strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >> irule le_mul >> simp[]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_MONO:
+    ∀f g s. FINITE s ∧ (∀x. x ∈ s ⇒ 0 ≤ f x ∧ f x ≤ g x) ⇒ ∏ f s ≤ ∏ g s
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >> irule le_mul2 >>
+    simp[EXTREAL_PROD_IMAGE_POS]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_COUNT_ZERO:
```

Comment:
`[simp]`

Reply:
This raises questions about:
`ITSET_EMPTY` (again)
`COUNT_ZERO`
`REAL_PROD_IMAGE_EMPTY`
`EXTREAL_PROD_IMAGE_EMPTY`
`REAL_SUM_IMAGE_EMPTY` (which doesn't exist, I could do that quickly)
`EXTREAL_SUM_IMAGE_EMPTY`
`EXTREAL_SUM_IMAGE_COUNT_ZERO`

Reply:
Done